DOC Restructure Multiclass and multilabel in User Guide

joshuacwnewton opened the pull request:
#### Reference Issues/PRs

Fixes #17728 

#### What does this implement/fix? Explain your changes.

Modifies [Section 1.12 Multiclass and multilabel algorithms](https://scikit-learn.org/stable/modules/multiclass.html#) as follows:

* Change section title to "Multiclass and multioutput algorithms" to match sklearn module names (`sklearn.multiclass` and `sklearn.multioutput`)
* Change overall section structure so that `headers == problem type`, and `subheaders == target format and meta-estimators`
* Redistribute content from introduction (problem types, target formats) into corresponding sections
* Rewrite the rest of the introduction to be concise, and to mention `sklearn.multioutput`
* Add chart image to introduction that shows hierarchy of topics/classes
* Fix `:mod:`/`:class:`/`:func:` links to `sklearn.multioutput`/`sklearn.utils.multiclass` that weren't rendering because of the presence of the `..currentmodule sklearn.multiclass` directive

#### Any other comments?

Apart from the introduction, right now the existing content is left largely unmodified -- I've simply moved it around into new header sections to improve clarity. That said, I feel some sections are still unclear, and could possibly use rewriting:

* The `Multiclass-multioutput classification` section is tricky. 
     * It's the combination of `multiclass` and `multioutput`, so it doesn't read like a distinct problem type. That makes it tough to fit into the topic hierarchy.
     * The existing content never mentions which meta-estimators support this. Is it the same as meta-estimators which support the `multilabel-indicator` target type? This could be clarified.
* `OneVsRestClassifier`, a multiclass estimator, can also do multilabel classification. This is unintuitive and potentially confusing! The class docstring was changed in #17646, but further changes could be made to the User Guide and API Reference to clarify the dual nature of this estimator.
* `ClassifierChain` and `RegressorChain` sections are somewhat brief, and don't have usage examples.

#### Regarding the chart figure

The chart image I added is possibly difficult to maintain. If changes are made to `sklearn.multiclass` or `sklearn.multioutput`, the image would need to be updated too. I wanted to generate the visualization programmatically, but I couldn't figure out how using Sphinx/matplotlb/etc. 

I created it using Google Drawings, so here is [a shareable link](https://docs.google.com/drawings/d/1roAkCbPbnbL4rR8KkSxvWCq9jVYQCBKbOJrz-ZgiH38/edit?usp=sharing) for anyone coming from the future wondering what the source of the chart image is. You should be able to make a copy and edit it if need be. :slightly_smiling_face:

joshuacwnewton committed:
Remove multilabel target format section

This information is already presented in the prelude of this User Guide
page.

joshuacwnewton committed:
Restructure headers to match problem domains

Previously, headers were organized by strategy type. However, I feel it
is easier to parse when organized by problem type. Strategies are
relegated to subheaders within each problem header. As well, a
subheader for target format is added.

No content is edited in this commit -- only headers.

joshuacwnewton committed:
Redistribute content from intro to new sections

Again, no edits made to the content itself -- simply redistributing
content from the introduction to the body sections of the page.

joshuacwnewton committed:
Retitle section "Multiclass and multioutput"

The previous title was "Multiclass and multilabel", but multilabel
doesn't account for multioutput regression. This new title is more
inclusive of all multilearning problem types.

joshuacwnewton committed:
Rewrite introduction to include multioutput module

Also included is a chart which outlines the responsibilities and
classes for each module.

joshuacwnewton committed:
Clarify multioutput/multilabel classification term

joshuacwnewton committed:
Fix class/function links

The ..currentmodule directive was set to `sklearn.multiclass`, so Sphinx
assumed all links were from that module. But, this guide contains links
to `sklearn.multioutput` and `sklearn.utils.multiclass`. To make them
render properly, I've removed the ..currentmodule directive, and updated
links accordingly.

joshuacwnewton committed:
Update subheader titles to class names

Subheaders were previously titled either using the name of the strategy
or the name of the problem type. When headers (not subheaders) were
changed to problem types, this resulted in some subheaders having the
same name as headers (e.g. "Multioutput Regression"). This change fixes
the conflicts.

joshuacwnewton committed:
Tweak multiclass-multioutput content

* Re-add entry to summary table
* Make sure all terms use "multiclass-multioutput" and not
  "multioutput-multiclass"

glemaitre commented:
@lucyleeow Since you invested some time to make an initial improvement on this section, it could be good if you could have a look.

lucyleeow reviewed (commented):
Thanks @joshuacwnewton, I like the new format and the diagram is helpful. I just did a quick pass.

> Remove subsection "1.12.1. Multilabel classification format". (It doesn't fit with the other subsections, which are devoted to different multiclass/multioutput strategies. Target formats/types are better expressed elsewhere.)

This removed subsection might be better put with section 6.9.1 on 'LabelBinarizer' [here](https://scikit-learn.org/stable/modules/preprocessing_targets.html). I think this can be changed in this PR since we are deleting it from here.

I'm not sure about putting 'multilabel' and 'multioutput' together. From the glossary:

> multilabel
>    A multiclass multioutput target where each output is binary. 

whereas:

> multi-output
>    A target where each sample has multiple classification/regression labels.

To me 'multioutput' is more general, it just means several labels, but each label does NOT have to binary.

> The existing content never mentions which meta-estimators support this. Is it the same as meta-estimators which support the multilabel-indicator target type? This could be clarified.

I think that multioutput meta-estimators, used with a 'base estimator' that support multiclass would work, but @glemaitre can confirm.

Another nitpick - since we are doing an over-haul, can I suggest we standardise how we write 'mutli<class, label etc>', either with a hypen (e.g., multi-label) or without (mutlilabel). We change between the 2 here.

joshuacwnewton commented:
Thanks much @glemaitre for connecting. And, thank you for your comments @lucyleeow. It is very nice to hear another perspective on this, because of how the definitions can sometimes be vague/ambiguous/contentious. 

Would you like to work together on this? There is sklearn's Gitter if it would be helpful for less formal brainstorming/chats about how to structure this section. Otherwise, here is good too. :) 

-----

>This removed subsection might be better put with section 6.9.1 on 'LabelBinarizer' here. I think this can be changed in this PR since we are deleting it from here.

Thank you for pointing this out. I didn't realize this section existed! I could also add a reference to this section in [`1.12.2.1 Target format`](https://110809-843222-gh.circle-artifacts.com/0/doc/modules/multiclass.html#id3) so that users can also go there to read more information, perhaps.

-----

> I'm not sure about putting 'multilabel' and 'multioutput' together. 

I very much agree when you say `multioutput` is the more general category, and `multilabel` is a subset. The challenge I was facing is the usage of these terms in the general literature, outside of sklearn.

`"multilabel classification"` seems to be used very consistently (e.g. [in survey/review papers](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=multilabel+classification&btnG=&oq=multilabel+clas)). But, `"multioutput classification"` doesn't show up nearly as much [in publication titles](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=multioutput+classification&btnG=&oq=multioutput+cl). See also the title of [the relevant Wikipedia page](https://en.wikipedia.org/wiki/Multi-label_classification), which newcomers would also likely refer to when trying to understand.

So, while sklearn does have a consistent set of definitions, I think in practice users would be searching for "multilabel classification" which makes things a bit difficult for us. I think it is good to make the association somehow, so that people can be pointed in the right direction, e.g. towards `MultiOutputClassifier`. But, perhaps there's a better way to make that connection?

----

>Another nitpick - since we are doing an over-haul, can I suggest we standardise how we write 'mutli<class, label etc>', either with a hypen (e.g., multi-label) or without (mutlilabel). We change between the 2 here.

I appreciate the suggestion, and very much agree that it is a good time for small changes like this. I was feeling hesitant to make too many changes at first, because I was just proposing, but now that you've shared your input, I feel more comfortable going ahead and standardising language like this. :)

----

>and link to the glossary, for each of these terms

Thank you for pointing this out. Like the above, I am very much in favour of small changes/standardisation like this. :)

lucyleeow commented:
> I could also add a reference to this section in 1.12.2.1 Target format so that users can also go there to read more information, perhaps.

Yes, I think it would be useful to refer to the `LabelBinarizer` guide here.

> The challenge I was facing is the usage of these terms in the general literature, outside of sklearn.

I think we should stick with the term 'mutli-label', which was used in the docs currently but maybe we can make it more explicit in the text that you can do 'multi-label' type problems using the `MultiOutputClassifier` class. I think adding the 'multi-output' to the titles is confusing and not entirely correct.

> I appreciate the suggestion, and very much agree that it is a good time for small changes like this.

Since we are making large changes to the 'blame' we may as well!

joshuacwnewton commented:
> I think we should stick with the term 'mutli-label', which was used in the docs currently but maybe we can make it more explicit in the text that you can do 'multi-label' type problems using the MultiOutputClassifier class. I think adding the 'multi-output' to the titles is confusing and not entirely correct.

I think this is a great suggestion. Also, if I update the chart as follows: 

![multiclass diagram](https://user-images.githubusercontent.com/16181459/87210945-3f56ee00-c2cc-11ea-9acb-b0cca20b9429.png)

Then the association between multilabel/multioutput is thankfully still made, because visually "Multilabel classification" appears underneath the `sklearn.multioutput` module.

joshuacwnewton commented:
Re: hyphen/no hyphen, it seems like it may be tough to ensure consistency all throughout the documentation. I see the differences everywhere now that I'm looking for it, hehe. Perhaps it is like `'` vs. `"` in Python strings. :stuck_out_tongue: 

Looking at the Glossary, it seems that both variations were added for [multi-output](https://scikit-learn.org/stable/glossary.html#term-multioutput). Perhaps it's best to simply amend the glossary to have both `multi-label` and `multilabel` so references link correctly regardless of personal preference.

joshuacwnewton committed:
Expand preprocessing_targets.rst for multilabel

* Move MultiLabelBinarizer description to preprocessing_targets.rst
* Add subheaders to separate multiclass/multilabel classification

joshuacwnewton committed:
Fix multiclass target example to have >2 classes

joshuacwnewton committed:
Refactor dense/sparse binary target example

Dense binary matrix was missing from multiclass target format section.
This also commit makes sure all examples in the multiclass target format
section use the same set of labels

joshuacwnewton committed:
Add cross-references for multiclass and binarizers

joshuacwnewton committed:
Change "Multioutput (multilabel) -> "Multilabel"

joshuacwnewton committed:
Add glossary links for terms in introduction

joshuacwnewton committed:
Update multilearning terms to include hyphen

thomasjpfan commented on doc/modules/multiclass.rst:
```diff
@@ -440,47 +334,66 @@ Below is an example of multiclass learning using Output-Codes::
+=========================
+
+**Multilabel classification** (closely related to **multioutput**
+**classification**) is a classification task labelling each sample with ``x``
```
```suggestion
**classification**) is a classification task labeling each sample with ``x``
```

joshuacwnewton replied:
There were a couple of instances of this -- I think I've caught them all. Thanks much! :)

thomasjpfan commented on doc/modules/multiclass.rst:
```diff
@@ -440,47 +334,66 @@ Below is an example of multiclass learning using Output-Codes::
+
+**Multilabel classification** (closely related to **multioutput**
+**classification**) is a classification task labelling each sample with ``x``
+labels from ``n_classes`` possible classes, where ``x`` can be 0 to
```
Using `x` in this context can get confused with the feature space i.e (`X`).

joshuacwnewton replied:
Good point! Addressed. :)

thomasjpfan reviewed (commented):
Thank you for the PR @joshuacwnewton 

Quick overview for now. Overall, the restructure looks good.

joshuacwnewton committed:
Fix 'labelled'/'labelling' typos

joshuacwnewton committed:
Replace ``x`` with ``m`` for clarity

joshuacwnewton commented:
Thanks much for taking a look, @thomasjpfan. No rush of course as this seems low-urgency, but just let me know, and I'm happy to come back to this when further reviews/changes are needed. :)

thomasjpfan commented on doc/modules/multiclass.rst:
```diff
@@ -1,181 +1,53 @@
+|                              | Number of targets     | Target cardinality      | Valid                                            |
+|                              |                       |                         | :func:`~sklearn.utils.multiclass.type_of_target` |
++==============================+=======================+=========================+==================================================+
+| Multiclass                   |  1                    | >2                      | - 'multiclass'                                   |
```
I think this looks slightly better without the `-`. (when rendered)

```suggestion
| Multiclass                   |  1                    | >2                      | 'multiclass'                                     |
```

thomasjpfan commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -12,7 +12,10 @@ you want to transform the prediction target for learning, but evaluate the
-------------------
+==================
+
+Multiclass classification
+-------------------------
```
sklearn considers this `'multilabel-indicator'`, i wonder if calling it "multiclass" classification would be confusing.

```py
from sklearn.utils.multiclass import type_of_target
import numpy as np
X = np.array([[1, 0, 0, 0],
              [0, 0, 0, 1]])
type_of_target(X)
# 'multilabel-indicator'
```

joshuacwnewton replied:
You're right, that is confusing. 

I wonder if that's more a limitation of `type_of_target`, though, as `LabelBinarizer` does seem to be [meant for multiclass classification](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html).

thomasjpfan replied:
It can be used for multiclass if its One-vs-Rest. One can consider this `"mutlilabel-indicator"` where there is only one label per sample.

I think saying that this is used for one-vs-rest multiclasss classification would make it a little more clear.

joshuacwnewton replied:
I've created #18145 to discuss this further. It might be overkill, but it could be useful for `type_of_target` to return something more specific in this case. :)

thomasjpfan reviewed (commented):
I think this is a good update to the user guide!

thomasjpfan reviewed (approved):
LGTM

glemaitre reviewed (approved):
LGTM as well. Before merging I was wondering if @lucyleeow has any comments since that she reviewed at the beginning?

lucyleeow commented on doc/modules/multiclass.rst:
```diff
@@ -440,47 +334,66 @@ Below is an example of multiclass learning using Output-Codes::
       Hastie T., Tibshirani R., Friedman J., page 606 (second-edition)
       2008.
 
```
Let's add a reference label:
`.. _multilabel_classification:`

lucyleeow commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -27,17 +30,38 @@ matrix from a list of multi-class labels::
+Multilabel classification
+-------------------------
+
+In multilabel learning, the joint set of binary classification tasks is
```
With our reference label, (above)
```suggestion
In :ref:`multilabel_classification`, the joint set of binary classification tasks is
```

lucyleeow commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -27,17 +30,38 @@ matrix from a list of multi-class labels::
+
+In multilabel learning, the joint set of binary classification tasks is
+expressed with label binary indicator array: each sample is one row of a 2d
+array of shape (n_samples, n_classes) with binary values: the one, i.e. the non
+zero elements, corresponds to the subset of labels. An array such as
```
nitpick
```suggestion
array of shape (n_samples, n_classes) with binary values where the one, i.e. the non
zero elements, corresponds to the subset of labels for that sample. An array such as
```

lucyleeow commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -27,17 +30,38 @@ matrix from a list of multi-class labels::
+           [1, 1, 1, 0, 0]])
+
+For more information about multilabel classification, refer to
+:ref:`multiclass`.
```
```suggestion
:ref:`multilabel_classification`.
```

lucyleeow reviewed (commented):
Just some suggestions for polishing off `reprocessing_targets.rst` but otherwise LGTM! It is great improvement!

joshuacwnewton committed:
Remove bullet list to improve table rendering

joshuacwnewton committed:
Clarify multiclass/multilabel in label binarization

The assertion that LabelBinarizer == multiclass classification
was incorrect. This commit clarifies the connection. A typo
is also fixed.

joshuacwnewton committed:
Add Sphinx reference labels to sections in multiclass.rst

joshuacwnewton committed:
Add additional multilearning :ref: and :term: markers

joshuacwnewton committed:
Address nitpick in preprocessing_targets.rst

joshuacwnewton committed:
Merge remote-tracking branch 'upstream/master' into 17728-restructure-multi-docs

joshuacwnewton commented on doc/modules/multiclass.rst:
```diff
@@ -241,53 +113,84 @@ if you're using one of these, unless you want custom multiclass behavior:
   - :class:`sklearn.ensemble.RandomForestClassifier`
 
 
```
I've noticed that there were other missing reference labels for other sections and estimators, so I've added them all to be thorough. Thanks for pointing this out! :smile:

joshuacwnewton commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -27,17 +30,38 @@ matrix from a list of multi-class labels::
-           [0, 0, 1]])
-    >>> lb.classes_
-    array([1, 2, 3])
+Multilabel classification
```
Considering that LabelBinarizer is repurposing the multilabel format in a way that it may not have been intended for, I've included the message using the `. warning::` directive, which I hope makes it clearer to the user that the usage is worth paying attention to.

If it's not too much trouble, could you check the new message in 0d16ce6a2f11d2fca92e96db690a6d2a26d3664a? I think my recommendations are appropriate, but given the topic, I could very well still be misunderstanding. :p

joshuacwnewton commented:
Thank you @thomasjpfan, @glemaitre, and especially @lucyleeow for your thoughtful and thorough reviews. :slightly_smiling_face:

jnothman commented on doc/modules/multiclass.rst:
```diff
@@ -240,54 +112,86 @@ if you're using one of these, unless you want custom multiclass behavior:
+      >>> print(y)
+      ['apple' 'pear' 'apple' 'orange']
+
+  - Dense or sparse :term:`binary` matrix of shape ``(n_samples, n_classes)``
```
I don't think this is a multiclass representation. I think this is a treatment of a multiclass problem as multiple binary problems. I.e. it is a recoding of the multiclass representation.

joshuacwnewton replied:
Ah, right! Yes. This was written when I still had the misconception.

You cleared this up for me in #18145, but I didn't update this section afterwards, so thank you for pointing this out.

joshuacwnewton replied:
I'm having trouble balancing being technically thorough with my language, yet still making things accessible to someone who may be new to these concepts. Which details to include, and which to leave out... hmm...

How about still mentioning this representation, but directing users to the `preprocessing_targets.rst` documentation instead?

```
.. warning::

    Certain estimators support binary data only. To get around this limitation,
    multiclass data can instead be represented using a :term:`label indicator matrix` 
    (which is typically used for multilabel classification). To transform multiclass data
    into this form, use :class:`~sklearn.preprocessing.LabelBinarizer`. For more 
    information, please refer to :ref:`preprocessing_targets`.
```

jnothman reviewed (commented):
Sorry I've not yet found time to look at more

lucyleeow commented on doc/modules/classes.rst:
```diff
@@ -1218,7 +1218,7 @@ Model validation
    :no-inherited-members:
 
-**User guide:** See the :ref:`multiclass` section for further details.
+**User guide:** See the :ref:`multiclass_classification` section for further details.
```
Would the ref to 'multilabel_classification' be better here?

joshuacwnewton replied:
From what I understood, multilabel in `sklearn.multiclass` is a relic from older sklearn versions. And, that we wanted to encourage users to use `sklearn.multioutput` for multilabel instead.

I think the confusion here is the section titles... `multiclass.py` shouldn't have "Multilabel classification" in the title anymore, since that is now handled by `multioutput.py`. 

I've made a change in 2356c8ef206b56d995f52b8b676c3315e513dbbc. But, it's possible I'm off-base here with my understanding.

joshuacwnewton replied:
See also https://github.com/scikit-learn/scikit-learn/pull/17811#issuecomment-656940275 for the suggestion that `multilabel` should be associated with `sklearn.multioutput`.

lucyleeow replied:
Hmm good point. We do talk about multi-label under the 'Multiclass classification strategies' section though:

> This method returns probabilities of class membership in both the single label and multilabel case. Note that in the multilabel case, probabilities are the marginal probability that a given sample falls in the given class. As such, in the multilabel case the sum of these probabilities over all possible labels for a given sample will not sum to unity, as they do in the single label case.

joshuacwnewton replied:
> OvR seems to support multilabel, though the other 2 don't.

You're right about that. 

There is related discussion in this issue https://github.com/scikit-learn/scikit-learn/issues/9602 (see https://github.com/scikit-learn/scikit-learn/issues/9602#issuecomment-323948432 in particular) that floats the idea of discouraging and/or deprecating OvR for multilabel. 

Personally, I like the idea of leaving the functionality in OvR for backwards compatibility but using documentation to nudge users towards `sklearn.multioutput` for their multilabel needs. But, that is a bigger discussion and I am a small contributor.

lucyleeow replied:
Makes sense! In that case leave it as is!

joshuacwnewton replied:
It might be good to hear from @jnothman as well considering I've quoted an older comment of his and it might not reflect his current feelings. 

e.g. If we really want to more clearly discourage `sklearn.multiclass.OneVsRestClassifier` for multilabel, there could be additional edits made to make it more explicit that `sklearn.multioutput` is the way to go instead.

jnothman replied:
I don't think it's this PR's job to remove multilabel support from OvR... Multilabel support in OvR is used internally to make binary estimators support multilabel automagically. It's a bit of a hack.

I still think we should be encouraging MultiOutputClassifier for multilabel cases, since OvR is not actually doing one-vs-rest in the multilabel case. MultiOutputClassifier also has a different probability API.

I think it would be good to discuss this with the core dev team and resolve if it is an issue worth breaking people's code for.

joshuacwnewton replied:
> I don't think it's this PR's job to remove multilabel support from OvR.

My apologies, I wasn't clear about what I was asking. I was just curious re: your thoughts on documentation _("What should we be encouraging?")_ and not about code. So, what you write here:

> I still think we should be encouraging MultiOutputClassifier for multilabel cases

Is the part that clarifies things for me. Thank you. :)

lucyleeow commented on doc/modules/preprocessing_targets.rst:
```diff
@@ -27,17 +30,49 @@ matrix from a list of multi-class labels::
+    repurposes this format for multiclass data to enable multiclass
+    classification in estimators that only support binary data.
+    Depending on your situation, you could consider using the
+    :term:`multiclass` target representation instead for estimators which
+    support other types of data.
```
Hmm I'm not completely clear on what you are saying here. Are you just suggesting to keep data in multiclass format when the estimator supports multiclass classification? If so maybe be more explicit, on first reading as it sounds a little like we are suggesting to transform to multiclass instead but we are saying they don't need to transform (as label binarizer is multiclass -> multilabel, so data would be multiclass to start with)

joshuacwnewton replied:
> Are you just suggesting to keep data in multiclass format when the estimator supports multiclass classification? If so maybe be more explicit, on first reading as it sounds a little like we are suggesting to transform to multiclass instead but we are saying they don't need to transform (as label binarizer is multiclass -> multilabel, so data would be multiclass to start with)

Thank you, this is exactly it. You've worded things in a much clearer way here.

cmarmo commented:
Hi @joshuacwnewton , will you be able to address comments and fix conflicts? Thanks!

joshuacwnewton commented:
Thanks much for the reminder, @cmarmo! I'll have it done by end of day today. :)

joshuacwnewton committed:
Merge remote-tracking branch 'upstream/master' into 17728-restructure-multi-docs

Resolve conflicts from PR 17820. That PR shortened links in multiclass.rst.
This merge preserves those shortened links.

joshuacwnewton committed:
Clarify LabelBinarizer warning to be more explicit

joshuacwnewton committed:
Remove "multilabel" from multiclass.py doc titles

Multilabel classification is now handled primarily by
`sklearn.multioutput`.

joshuacwnewton committed:
classes.rst: Restore accidentally removed lines

I made a mistake while resolving the recent merge conflict.

cmarmo commented:
@thomasjpfan, two approvals here? Time to merge? Thanks!

thomasjpfan commented:
Thank you for the ping @cmarmo . Yup lets merge this.

thomasjpfan commented:
Thank you for working on this @joshuacwnewton !
